linux/tests/async_socket_linux_int/async_socket_linux_int.c

```diff
@@ -0,0 +1,520 @@
+TEST_DEFINE_ENUM_TYPE(ASYNC_SOCKET_SEND_SYNC_RESULT, ASYNC_SOCKET_SEND_SYNC_RESULT_VALUES)
+TEST_DEFINE_ENUM_TYPE(ASYNC_SOCKET_SEND_RESULT, ASYNC_SOCKET_SEND_RESULT_VALUES)
+
+static void on_open_complete(void* context, ASYNC_SOCKET_OPEN_RESULT open_result)
+{
+    volatile_atomic int32_t* thread_counter = (volatile_atomic int32_t*)context;
+
+    ASSERT_ARE_EQUAL(ASYNC_SOCKET_OPEN_RESULT, ASYNC_SOCKET_OPEN_OK, open_result);
+    (void)interlocked_increment(thread_counter);
+    wake_by_address_single(thread_counter);
+}
+
+static void on_send_complete(void* context, ASYNC_SOCKET_SEND_RESULT send_result)
+{
+    volatile_atomic int32_t* thread_counter = (volatile_atomic int32_t*)context;
+
+    ASSERT_ARE_EQUAL(ASYNC_SOCKET_SEND_RESULT, ASYNC_SOCKET_SEND_OK, send_result);
+    (void)interlocked_increment(thread_counter);
+    wake_by_address_single(thread_counter);
+}
+
+static void on_receive_complete(void* context, ASYNC_SOCKET_RECEIVE_RESULT receive_result, uint32_t bytes_received)
+{
+    volatile_atomic int32_t* thread_counter = (volatile_atomic int32_t*)context;
+
+    ASSERT_ARE_EQUAL(ASYNC_SOCKET_RECEIVE_RESULT, ASYNC_SOCKET_RECEIVE_OK, receive_result);
+    (void)interlocked_increment(thread_counter);
+    wake_by_address_single(thread_counter);
+}
+
+static void on_receive_and_accumulate_complete(void* context, ASYNC_SOCKET_RECEIVE_RESULT receive_result, uint32_t bytes_received)
+{
+    volatile_atomic int32_t* accumulator = (volatile_atomic int32_t*)context;
+
+    ASSERT_ARE_EQUAL(ASYNC_SOCKET_RECEIVE_RESULT, ASYNC_SOCKET_RECEIVE_OK, receive_result);
+    (void)interlocked_add(accumulator, bytes_received);
+    wake_by_address_single(accumulator);
+}
+
+static void on_receive_complete_with_error(void* context, ASYNC_SOCKET_RECEIVE_RESULT receive_result, uint32_t bytes_received)
+{
+    volatile_atomic int32_t* thread_counter = (volatile_atomic int32_t*)context;
+
+    ASSERT_ARE_EQUAL(ASYNC_SOCKET_RECEIVE_RESULT, ASYNC_SOCKET_RECEIVE_ABANDONED, receive_result);
+    (void)interlocked_increment(thread_counter);
+    wake_by_address_single(thread_counter);
+}
+
+static void set_nonblocking(SOCKET_HANDLE socket)
+{
+    int opts = fcntl(socket, F_GETFL);
+    ASSERT_IS_TRUE(opts >= 0, "Failure getting socket option");
+
+    opts = fcntl(socket, F_SETFL, opts|O_NONBLOCK);
+    ASSERT_IS_TRUE(opts >= 0, "Failure setting socket option");
+}
+
+static void setup_server_socket(SOCKET_HANDLE* listen_socket)
+{
+    // create a listening socket
+    *listen_socket = socket(AF_INET, SOCK_STREAM, 0);
+    ASSERT_ARE_NOT_EQUAL(int, INVALID_SOCKET, *listen_socket);
+
+    const int enable = 1;
+    ASSERT_ARE_EQUAL(int, 0, setsockopt(*listen_socket, SOL_SOCKET, SO_REUSEADDR, &enable, sizeof(int)));
+
+    int bind_res;
+    uint32_t counter = 0;
+    // The gate machines fails on bind due to 3 different
+    // Process running, so we have to figure out the bind
```

Comment:
>figure out the bind [](http://example.com/codeflow?start=38&length=19)

I think we should state that we're running through a set of ports and picking the first one that succeeds